File: backend/app/schemas/agent_runs.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, JsonValue, field_validator

from app.agents.schemas import RejectedChunk, VerifiedChunk


class AgentRunEvidenceResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    verified_chunks: list[VerifiedChunk]
    rejected_chunks: list[RejectedChunk]


class AgentRunLogStepResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    step_name: str = Field(min_length=1)
    agent_name: str = Field(min_length=1)
    input: dict[str, JsonValue]
    output: dict[str, JsonValue]
    status: Literal["success", "failed"]
    created_at: datetime
    error_message: str | None

    @field_validator("step_name", "agent_name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("name must not be empty")
        return normalized


class AgentRunLogsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    agent_run_id: UUID
    steps: list[AgentRunLogStepResponse]


__all__ = [
    "AgentRunEvidenceResponse",
    "AgentRunLogsResponse",
    "AgentRunLogStepResponse",
]

File: frontend/src/types/agentRuns.ts
export const AGENT_STEP_NAMES = {
  retrieval: "agent_1_retrieval",
  verification: "agent_2_verification",
  answerSelfCheck: "agent_3_answer_self_check",
} as const;

export type RecognizedAgentStepName =
  (typeof AGENT_STEP_NAMES)[keyof typeof AGENT_STEP_NAMES];

export type AgentStepName =
  | RecognizedAgentStepName
  | (string & Record<never, never>);

export type AgentStepStatus = "success" | "failed";

export type AgentStep = {
  agent_name: string;
  step_name: AgentStepName;
  input: unknown;
  output: unknown;
  status: AgentStepStatus;
  created_at: string;
  error_message: string | null;
};

export type AgentRunLogsResponse = {
  agent_run_id: string;
  steps: AgentStep[];
};
